framework/source/class/qx/event/Utils.js

```diff
@@ -0,0 +1,328 @@
+        var result = fn();
+        if (result === qx.event.Utils.ABORT) {
+          return this.reject(tracker);
+        }
+        return result;
+      }
+    }),
+    
+    /**
+     * Helper method to append a promise after the current one
+     * 
+     * @param tracker {Object} the tracker object
+     * @param newPromise {qx.Promise} the new promise
+     * @return {qx.Promise} the new promise
+     */
+    __thenPromise: function(tracker, newPromise) {
+      if (tracker.promise) {
+        this.__push(tracker, tracker.promise.then(function() { 
+            return newPromise; 
+          }));
+      } else {
+        this.__push(tracker, newPromise);
+      }
+      this.__addCatcher(tracker);
+      return tracker.promise;
+    },
+    
+    /**
+     * Rejects the tracker, aborting the promise if there is one.  The caller should stop 
+     * immediately because if promises are not in use and exception is not thrown.
+     * 
+     * @param tracker {Object} the tracker object
+     * @return {qx.Promise?} the last promise or the value returned by the catcher
+     */
+    reject: function(tracker) {
+      if (tracker.rejected) {
+        return qx.event.Utils.ABORT;
+      }
+      tracker.rejected = true;
+      
+      if (tracker.promise) {
+        throw new Error("Rejecting Event"); 
+      }
+      var result = this.__catcher(tracker);
+      return result === undefined ? this.ABORT : result;
+    },
+    
+    /**
+     * Helper method that adds a catcher to the tracker
+     * 
+     * @param tracker {Object} the tracker object
+     */
+    __addCatcher: function(tracker) {
+      if (tracker.promise && tracker.catch) {
+        if (!tracker.promise["qx.event.Utils.hasCatcher"]) {
+          this.__push(tracker, tracker.promise.catch(this.__catcher.bind(this, tracker)));
+          tracker.promise["qx.event.Utils.hasCatcher"] = true;
+        }
+      }
+    },
+    
+    /**
+     * This method is added with `.catch` to every promise created; because this is added
+     * all the way up the promise chain to ensure that it catches everything, this method
+     * supresses multiple invocations (i.e. ignores everything except the first)
+     * 
+     * @param tracker {Object} the tracker object
+     */
+    __catcher: function(tracker, err) {
+      var fn = tracker.catch;
+      if (fn) {
+        tracker.catch = null;
+        tracker.rejected = true;
+        return fn(err);
+      }
+      return qx.event.Utils.ABORT;
+    },
+    
+    /**
+     * Equivalent to `.catch()`; note that unlike promises, this method must be called *before*
+     * `.then()` so that it is able to handle rejections when promises are not in use; this is 
+     * because `Promise.catch` only catches rejections from previous promises, but because promises 
+     * are *always* asynchronous the `.catch` goes at the end.  For synchronous, this is nt possible 
+     * so `Utils.catch` must go before `Utils.then` 
+     * 
```

Comment:
This similarity in names but difference in how it must be used is very likely to cause people problems. Maybe it should have a different name?

Reply:
I wondered about that, and I'm open to suggestions - the dilemma is that it is modelled after promises, which is conceptually convenient at times but not at others.  In many cases this class is only for use as an internal class, so what do you think about if the comments at the top should make that clearer?

Reply:
Are you able to trap, somehow, when `catch` is used in the "wrong" place, i.e., at the end instead of before `when`? If so, I think it's fine to name it this way along with that comment. If not... hmmm... rename `catch` to be `catchMustBeBeforeWhen`? Sigh.

Reply:
re trapping wrong catch: it's not easy because it can come down to the use case and it's preserving a synchronous model **until** promises are invoked and then it switches to asynchronous which turns the exception handling upside down.  I could trap that `catch` was called after the rejection, but that's not necessarily invalid - also, I'm cautious that we will start to "go down the rabbit hole" of creating some kind of well defined and generic AlmostButNotQuitePromise when what's really needed is just some helper methods that are abstract enough to enable unit testing.

I resisted calling the class `qx.event.FastPromise` or similar to avoid making it look like anything other than internal helper methods for specific tasks, we could go further with names like `qx.event.Utils.psuedoThen` and `qx.event.Utils.psuedoCatch`?  

But then how far do we go, given that it is impossible to create a fully promise compliant solution, especially if the code is more clearly marked as internal?

Reply:
Document the crap out of it, and let it be, I guess.